Add support for LT8491

jemfgeronimo opened the pull request:
## Pull Request Description

The LT8491 is a buck-boost switching regulator battery charger that implements a constant-current constant-voltage (CCCV) charging profile used for most battery types, including sealed lead-acid (SLA), flooded, gel and lithium-ion.

The device operates from input voltages above, below or equal to the output voltage and can be powered by a solar panel or a DC power supply. On-chip logic provides automatic maximum power point tracking (MPPT) for solar powered applications. The LT8491 can perform automatic temperature compensation by sensing an external thermistor thermally coupled to the battery. The STATUS pin can be used to drive an LED indicator lamp. The device is available in a low profile (0.75mm) 7mm x 11mm 64-lead QFN package.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

CiprianRegus commented on drivers/power/lt8491/lt8491.c:
```diff
@@ -0,0 +1,612 @@
+#include "no_os_util.h"
+#include <string.h>
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+/******************************************************************************/
```
Please remove these comments.

CiprianRegus commented on drivers/power/lt8491/lt8491.c:
```diff
@@ -0,0 +1,612 @@
+ * @param charval - char to store telemetry battery charging stage
+ * @return 0 in case of success, errno errors otherwise
+ */
+int lt8491_read_charging_stage(struct lt8491_device *dev, int32_t *charval)
```
You could just use the enum values in an application directly. Why is there a need to assign a character to them?

CiprianRegus commented on drivers/power/lt8491/lt8491.h:
```diff
@@ -0,0 +1,198 @@
+/**
+ * @brief LT8491 descriptor
+ */
+struct lt8491_device {
```
we usually name these `_desc`. Can you please do the same here?

CiprianRegus commented on drivers/power/lt8491/iio_lt8491.c:
```diff
@@ -0,0 +1,361 @@
+static int32_t lt8491_iio_reg_read(struct lt8491_iio_device *dev,
+				   uint32_t reg, uint32_t *readval)
+{
+	return lt8491_reg_read(dev->dev, (uint8_t)reg, (uint16_t*)readval,
```
You will have to use a `uint16_t *` variable in order to read the value. Otherwise the result will be different depending on CPU endianess.

CiprianRegus commented on drivers/power/lt8491/iio_lt8491.c:
```diff
@@ -0,0 +1,361 @@
+		return -EOPNOTSUPP;
+	}
+
+	if (ret)
```
You can remove this, as ret will always be 0.

CiprianRegus commented on drivers/power/lt8491/iio_lt8491.c:
```diff
@@ -0,0 +1,361 @@
+	return len;
+}
+
+static struct iio_attribute lt8491_iio_attrs[] = {
+	LT8491_IIO_CH_ATTR_RO("tbat", LT8491_TBAT),
+	LT8491_IIO_CH_ATTR_RO("pout", LT8491_POUT),
+	LT8491_IIO_CH_ATTR_RO("pin", LT8491_PIN),
+	LT8491_IIO_CH_ATTR_RO("efficiency", LT8491_EFF),
+	LT8491_IIO_CH_ATTR_RO("iout", LT8491_IOUT),
+	LT8491_IIO_CH_ATTR_RO("iin", LT8491_IIN),
+	LT8491_IIO_CH_ATTR_RO("vbat", LT8491_VBAT),
+	LT8491_IIO_CH_ATTR_RO("vin", LT8491_VIN),
+	LT8491_IIO_CH_ATTR_RO("charging_stage", LT8491_CHARGING_STAGE),
+	LT8491_IIO_CH_ATTR_RO("charging_status", LT8491_CHARGING_STATUS),
+	LT8491_IIO_CH_ATTR_RW("enable", LT8491_ENABLE),
+	LT8491_IIO_CH_ATTR_WO("reset", LT8491_RESTART),
+	LT8491_IIO_CH_ATTR_RW("scratch", LT8491_SCRATCH),
+	LT8491_IIO_CH_ATTR_RO("serial_id", LT8491_SERIAL_ID),
+	END_ATTRIBUTES_ARRAY
+};
```
These should all be debug attributes. "Standard" IIO attibutes are documented on the IIO ABI page https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-bus-iio and we should strive to use the same ones for no-OS too. Here is the list of attributes supported by no-os https://github.com/analogdevicesinc/no-OS/blob/main/iio/iio.c#L109 and a driver which uses both types https://github.com/analogdevicesinc/no-OS/blob/main/drivers/imu/iio_adis.c

jemfgeronimo commented:
V2
- moved global attributes to debug attributes
- some global attributes are now iio channel attributes
- renamed `lt8491_device` to `lt8491_desc`
- removed big comments, e.g., `/*Macros and Constants Definitions*/`
- modified `lt8491_read_charging_stage` to get `int` rather than `char`

kister-jimenez reviewed (commented):
Only minor comments that needs to be addressed. Other than that, looks good to me.

jemfgeronimo commented on projects/dc2703a/README.rst:
```diff
@@ -0,0 +1,151 @@
+Overview
+--------
+
+The DC2703A-A-KIT contains the DC2703A (LT8491 demo board) and DC1613A
+(USB-to-I2C controller). Together they provide a high performance buck-boost
+battery charger converter with an I2C interface. The LT8491 implements a maximum
+power point tracking (MPPT) function and flexible charging profiles, suitable
+for most battery types such as flooded and sealed lead  acid batteries and
+Li-Ion batteries.
```
Will this suffice?
```suggestion
The DC2703A-A-KIT houses the DC2703A (LT8491 demo board). They provide a high
performance buck-boost battery charger converter with an I2C interface for a
microcontroller host. The LT8491 implements a maximum power point tracking
(MPPT) function and flexible charging profiles, suitable for most battery types
such as flooded and sealed lead acid batteries and Li-Ion batteries.
```

CiprianRegus reviewed (dismissed):
No-OS has recently transitioned to a BSD 3 clause license (from ADI BSD). Please update the license headers for all files. Otherwise, it looks good.

The new license: https://github.com/analogdevicesinc/no-OS/blob/main/LICENSE_BSD

jemfgeronimo commented:
V3
- improved `Overview` in `projects/dc2703a/README.rst`
- fixed typo in `projects/dc2703a/src/platform/maxim/main.c`
- updated license headers to `BSD 3` clause license (from `ADI BSD`)
